Don't create sessions when connecting to etcd in the clustermesh context

giorio94 opened the pull request:
Currently, the etcd client leverages a session to wait for the establishment of the initial connection. However, this potentially introduces a quite significant overhead in the context of clustermesh, due to the usage of short-lived (25 seconds) etcd sessions which grow in number proportionally to the Cilium agents in each given cluster, times the number of clusters in the clustermesh. Yet, these sessions, despite being renewed until the client gets closed, are never used after this initial check. 

In an effort to improve the overall clustermesh performance, and reduce the overhead introduced on the etcd instances, let's rework this logic to not depend on the creation of a session in the clustermesh context (i.e., if the lock quorum check is disabled), but rather just wait until the heartbeat watcher is started successfully (i.e., the list operation completed). 

Most notably, this (a) preserves the current behavior of the etcd client reporting readiness only once successfully
connected, (b) guarantees that there's already been an interaction with the target etcd instance (even if the session check is disabled), to ensure that its corresponding cluster ID has been retrieved if using the "clusterLock" interceptors (i.e., in the clustermesh context) and (c) prevents the client from turning ready if the heartbeat watcher cannot be started, avoiding regressions in this respect and ensuring that the status checker logic can operate correctly.

giorio94 committed:
etcd: make waitForInitLock function synchronous

Convert the waitForInitLock function to run synchronously, rather than
internally starting a separate goroutine and returning the result via
a channel. This makes the overall logic more straightforward to reason
about, especially in case the lock quorum check is disabled (whose check
has been anticipated to the beginning of the function).

While being there, let's also rename the function to maybeWaitForInitLock,
to better explicit the fact that this check may be skipped if disabled
via the corresponding option.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
etcd: simplify hearbeat watcher setup logic

An etcd watcher is automatically closed upon context cancelling, hence
we don't need any special handling for that. Additionally, let's make
the events channel unbuffered, as there's no point in performing any
buffering in this context.

Finally, let's ignore deletion events, as they do not represent a
valid heartbeat signal. Differently, we consider a ListDone event
as a valid signal, to ensure that the lastHeartbeat variable is
set to a non-zero value, effectively starting the status check.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
etcd: don't create unnecessary sessions in the clustermesh context

Currently, the etcd client leverages a session to wait for the
establishment of the initial connection. However, this potentially
introduces a quite significant overhead in the context of clustermesh,
due to the usage of short-lived (25 seconds) etcd sessions which grow
in number proportionally to the Cilium agents in each given cluster,
times the number of clusters in the clustermesh. Yet, these sessions,
despite being renewed until the client gets closed, are never used
after this initial check.

In an effort to improve the overall clustermesh performance, and reduce
the overhead introduced on the etcd instances, let's rework this logic
to not depend on the creation of a session in the clustermesh context
(i.e., if the lock quorum check is disabled), but rather just wait
until the heartbeat watcher is started successfully (i.e., the list
operation completed). Most notably, this (a) preserves the current
behavior of the etcd client reporting readiness only once successfully
connected, (b) guarantees that there's already been an interaction with
the target etcd instance (even if the session check is disabled), to
ensure that its corresponding cluster ID has been retrieved if using
the "clusterLock" interceptors (i.e., in the clustermesh context) and
(c) prevents the client from turning ready if the heartbeat watcher
cannot be started, avoiding regressions in this respect and ensuring
that the status checker logic can operate correctly.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

marseel commented on pkg/kvstore/etcd.go:
```diff
@@ -656,30 +656,18 @@ func connectEtcdClient(ctx context.Context, config *client.Config, cfgPath strin
-				return
-			}
+		watcher := ec.ListAndWatch(ctx, HeartbeatPath, 0)
+		for range watcher.Events {
```
nit: Not related to this PR, but this seems like a bug. 
We should probably only care about update events, not delete/list done events.

giorio94 replied:
I think it depends, as even an unrelated event is still a liveness signal (at least of the watch stream).

Considering the current implementation, considering the list done event is actually useful to set the last heartbeat variable to a non-zero value, which then actually enables the check. Otherwise, it would be skipped due to:

https://github.com/cilium/cilium/blob/7e9de944a06eca20a88abd8d072d751dd3b56447/pkg/kvstore/etcd.go#L1034-L1037

I instead agree that a delete event could be instead safely ignored.

marseel replied:
:thinking: :thinking: :thinking: 
Sorry, I don't understand it. 

ListDone is an artificially created event [here](https://github.com/cilium/cilium/blob/f30d83a2476adaae74cdcf27a395e3c48d258399/pkg/kvstore/etcd.go#L872)

I would always assume that heartbeat events correspond to something that happened in relation to the outside world.

> Considering the current implementation, considering the list done event is actually useful to set the last heartbeat variable to a non-zero value, which then actually enables the check. Otherwise, it would be skipped due to:

these all checks are very suspicious. If we never create a watch and keep relisting the same key, it would still never fail, no?

imho we should be checking if heartbeat value changed since last time we got it and only then update `lastHeartbeat`.

giorio94 replied:
> ListDone is an artificially created event [here](https://github.com/cilium/cilium/blob/f30d83a2476adaae74cdcf27a395e3c48d258399/pkg/kvstore/etcd.go#L872)

Yes, my point is that it is a signal that we actually successfully listed the heartbeat key (which may, or may not exist at this time). Setting the `lastHeartbeat` variable in that case ensures that if we don't receive any event within the expected timeout, we'll consider that as an heartbeat failure, which looks the correct behavior to me. Especially considering that with the new changes the status checker is only started once receiving the list done event (which again looks correct to me, as we want to run the status checker only once exited the initial connection phase, as guarded by a separate timeout).

> these all checks are very suspicious. If we never create a watch and keep relisting the same key, it would still never fail, no?

Sort of. That's what happened when I accidentally dropped the permissions for the heartbeat prefix, preventing it from starting. This specific case is now prevented as the list operation needs to succeed before considering the client ready. There's definitely the possibility that the watch keeps subsequently failing, triggering list operation agains, and the corresponding upsert event, which would incorrectly count as an heartbeat. Although arguably that looks extremely rare to me, considering that it would require the watcher to continuously fail due to a compacted error.

That said, I totally agree with you that the status checker logic could be improved to address these corner cases and make the overall logic more straightforward to reason about. Although I'm always a bit scared of touching it due to the possible unforeseen consequences on Cilium running in kvstore mode, as the lifecycle there is completely different.

marseel replied:
> Yes, my point is that it is a signal that we actually successfully listed the heartbeat key (which may, or may not exist at this time). Setting the lastHeartbeat variable in that case ensures that if we don't receive any event within the expected timeout, we'll consider that as an heartbeat failure, which looks the correct behavior to me.

I agree this is correct behavior, but it strongly suggests an incorrect division of responsibilities between heartbeat watcher and statusChecker. 
In that case, we should set lastHeartbeat to time.now() when starting the status checker initially: https://github.com/cilium/cilium/blob/7e9de944a06eca20a88abd8d072d751dd3b56447/pkg/kvstore/etcd.go#L1031
Successful detection of connection problem should not depend on the fact if we receive any event or not. 
Does it make sense?

giorio94 replied:
Sure, makes sense. I'm going to update that.

marseel commented on pkg/kvstore/etcd.go:
```diff
@@ -407,44 +408,32 @@ func (e *etcdClient) StatusCheckErrors() <-chan error {
-				return
-			}
+func (e *etcdClient) waitForInitLock(ctx context.Context) error {
+	if e.extraOptions != nil && e.extraOptions.NoLockQuorumCheck {
```
nit: I would probably move this check up to isConnectedAndHasQuorum
or alternatively change the name to `maybeWaitForInitLock` to make it more explicit. 
I slightly prefer first option though. 

Maybe even inline this function now within `isConnectedAndHasQuorum` so it's not accidentally used somewhere else accidentally.

giorio94 replied:
Totally agree that this function is confusing. 

I initially thought about moving the check to `isConnectedAndHasQuorum`, but then decided against it in case anyone ended up calling `waitForInitLock` directly. I'll rename the function as you suggested to clarify it for the moment. Hopefully we can simplify this logic even more once dropping the support for etcd running in pod network, as mostly used to stop the informers during the handover process.

marseel commented on pkg/kvstore/etcd.go:
```diff
@@ -708,6 +703,20 @@ func makeSessionName(sessionPrefix string, opts *ExtraOptions) string {
 
+func (e *etcdClient) waitForReadyConnection(ctx context.Context) error {
+	for {
+		state := e.client.ActiveConnection().GetState()
```
hmm, it's very non-obvious to me why the connection wouldn't be in an IDLE state. It means that something somewhere had to issue an RPC call as per GRPC documentation:
```
IDLE: This is the state where the channel is not even trying to create a connection because of a lack of new or pending RPCs. New RPCs MAY be created in this state. Any attempt to start an RPC on the channel will push the channel out of this state to connecting. When there has been no RPC activity on a channel for a specified IDLE_TIMEOUT, i.e., no new or pending (active) RPCs for this period, channels that are READY or CONNECTING switch to IDLE. Additionally, channels that receive a GOAWAY when there are no active or pending RPCs should also switch to IDLE to avoid connection overload at servers that are attempting to shed connections. We will use a default IDLE_TIMEOUT of 300 seconds (5 minutes).
```
[source](https://grpc.github.io/grpc/core/md_doc_connectivity-semantics-and-api.html)
So essentially waiting here on Ready state AFAIK depends on this call in newClient AFAIK:
https://github.com/cilium/cilium/blob/09872a6758f52b40f9d540e076bbf1eef1100579/vendor/go.etcd.io/etcd/client/v3/client.go#L450

but then the question is, is it worth doing that if newClient already performs that call and would return an error if RPC failed?

giorio94 replied:
My understanding is that the connection transitions out of idle mode due to the `newClient` calling `dialWithBalancer`, which eventually calls `grpc.DialContext`:

https://github.com/cilium/cilium/blob/7e9de944a06eca20a88abd8d072d751dd3b56447/vendor/go.etcd.io/etcd/client/v3/client.go#L429

https://github.com/cilium/cilium/blob/7e9de944a06eca20a88abd8d072d751dd3b56447/vendor/go.etcd.io/etcd/client/v3/client.go#L318

According to the `DialContext`'s godoc:

https://github.com/cilium/cilium/blob/7e9de944a06eca20a88abd8d072d751dd3b56447/vendor/google.golang.org/grpc/clientconn.go#L209-L211

Given that the goal here is to wait for the connection to be established (with a timeout), one possibility could be to set the `WithBlock` dial option, and configure a non-zero DialTimeout. However, `WithBlock` is [deprecated and considered an anti-pattern](https://github.com/grpc/grpc-go/blob/master/Documentation/anti-patterns.md). The documentation suggests to use `WaitForStateChange` instead.

Differently, `getToken` should not have effects, because neither the username nor the password are set as we use certs for authentication:

https://github.com/cilium/cilium/blob/7e9de944a06eca20a88abd8d072d751dd3b56447/vendor/go.etcd.io/etcd/client/v3/client.go#L272-L277

I guess that yet another possibility (maybe the best one actually) could be removing the `waitForReadyConnection` step and just wait for the `ListDone` event on the heartbeat prefix. That would ensure that we are actually connected, and not depend on the underlying grpc connection status (which may indeed turn idle again after a timeout if no requests are issued). WDYT?

marseel replied:
yea, waiting on ListDone from heartbeat sounds like the most reasonable approach. I would prefer to not know about etcd client / grpc internals :sweat_smile:

giorio94 replied:
Yep, I agree. Let me change that and verify that it works as expected.

giorio94 replied:
I did a bit of testing, and it seemed to work great :rocket:

marseel reviewed (changes requested):
2 small non-blocking nits, one more interesting comment

giorio94 committed:
etcd: explicitly initialize lastHeartbeat in statusChecker

The lastHeatbeat value is updated upon reception of heartbeat events.
Currently, it is additionally set upon the list done event: while the
outcome is technically correct, leading it to be always initialized
before starting the statusChecker, it is also a code smell, as it
mixes responsibilities. Let's instead move the initialization to the
statusChecker function itself, and drop the subsequent IsZero check,
that already led to issues in the part.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

marseel commented on pkg/kvstore/etcd.go:
```diff
@@ -612,91 +597,103 @@ func connectEtcdClient(ctx context.Context, config *client.Config, cfgPath strin
-			close(ec.firstSession)
-			close(ec.statusCheckErrors)
-			return
+	if err != nil {
```
nit: this could be withing if block above for easier reading.

giorio94 replied:
Yep, makes sense. Let me fix it.

marseel reviewed (approved):
Looks good to me, thanks!
